Add screen hierarchy for spectating another player

peppy opened the pull request:
Just going to PR this to get something out there (and so I can monitor the size of the diff)

Note that there's no way to reach this from the client itself, and I've been developing it completely via visual tests. Individual commits can be split out if deemed required. A few underlying changes were required to make this work:

- `FrameStabilityContainer` needed logic changes to handle cases where replay data was not completely received. Fixing handling where there's zero frames, or where we've reached the end of streaming data.
- Player needs to act on `FrameStabilityContainer` running out of frames, and pause gameplay. Right now this is done via a simple method, which will likely need to be adjusted once there's the ability for a user to pause replay playback.

Note that some of the tests are incomplete (pending further implementation in the `Spectator` screen component). All the basic flows and behaviours seem to be correct though.

peppy committed:
Add basic spectator screen

peppy committed:
Expose events from streaming client

peppy committed:
Add basic implementation of spectator screen

peppy committed:
Create basic testing setup

peppy committed:
Convey actual beatmap and ruleset for full testing setup

peppy committed:
Setup tests

peppy committed:
Setup tests to run headless, add basic pass support

peppy committed:
Ensure frames arrive

peppy committed:
Add null check on replay as safety measure

peppy committed:
Stop replay playback when frames are starved

peppy committed:
Add field to Replay denoting whether the full replay have been received or not

peppy committed:
Add more useful frame sending logic to tests

peppy committed:
Allow FrameStabilityContainer to handle waiting-for-data state better (and pause outwards)

peppy committed:
Send initial spectator state more correctly in test component

peppy committed:
Add missing space

peppy committed:
Add sane pausing logic

peppy committed:
Allow tests to continue sending frames from point they left off

peppy committed:
Fix gameplay proceeding when no frames have been received yet

peppy committed:
Add test helper functions to promote code share

peppy committed:
Fix spectating when starting from a point that isn't at the beginning of the beatmap

peppy committed:
Add basic beatmap download and play flow

peppy committed:
Fix a couple of inspections

bdach reviewed (dismissed):
Here's a rudimentary first pass. Haven't looked at this too hard, especially given that tests are yet to pass.

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneSpectator.cs:
```diff
@@ -0,0 +1,236 @@
+
+            AddStep("reset sent frames", () => nextFrame = 0);
+
+            AddStep("import beatmap", () => ImportBeatmapTest.LoadOszIntoOsu(game, virtualTrack: true).Wait());
```
yeah, was planning on replacing this with a randomised beatmap to avoid weird edge cases. the import is very temporary.

peppy commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -124,22 +124,50 @@ private bool inImportantSection
+                        return null;
+                    }
+
+                    // if we didn't change frames, we need to ensure we are allowed to run frames in between, else return null.
```
nah, that comment is talking about the fall-through case.

and i completely agree that this method is shocking, but i spent a good hour looking at it and didn't come up with anything better. will take another look at it today, but open to any improvements you can see, even if they are minor steps in the right direction.

curious if you can repro that behaviour on master?

peppy replied:
Rewrote this logic, hopefully the issue is just-resolved now (I still can't see this happening locally).

peppy commented:
Funny part is all tests are/were passing 100% locally 😅. Always the way.

peppy committed:
Avoid null reference on finalization

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

bdach commented:
Yeah, I'm not even 100% sure what that `Push` failure is all about. Looks like perhaps the spectator client might be calling `Push` too eagerly? Cannot repro locally, so mostly guessing...

peppy committed:
Make direction setting more clear

peppy commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -124,22 +124,50 @@ private bool inImportantSection
+                    else if (next.Time.CompareTo(time) != currentDirection)
+                    {
+                        // in the case we have more frames, block if the next frame's time is less than the current time.
+                        return null;
```
I've added this outside the if-else clause as it's `!= 0` in all cases.

peppy committed:
Merge branch 'frame-stability-clean-up' into spectator-replay-watcher

peppy committed:
Merge branch 'frame-stability-clean-up' into spectator-replay-watcher

peppy committed:
Simplify FramedReplayInputHandler's SetFrame implementation

peppy committed:
Fix "finished playing" events handled for potentially incorrect user

peppy committed:
Ensure spectator screen is loaded before continuing

peppy committed:
Avoid reconstructing ruleset for each frame bundle

peppy committed:
Fix switching to new beatmap not working correctly

peppy committed:
Construct replay after being sure a ruleset is available to avoid nullrefs

peppy committed:
Fix attempt to change ruleset/beatmap bindables while screen is not active

peppy committed:
Always use imported beatmap in tests

peppy committed:
Change beatmap not existing test to specify a beatmap ID that can't possibly exist

peppy committed:
Add more accurate fail scenario test logic

peppy committed:
Avoid cross-pollution between tests of current playing state

bdach commented:
I think I've understood why I'm getting this weird oscillation issue. [Here's a demonstration video of how it looks to me.](https://drive.google.com/file/d/1EMPuogknGkJu0tBUjM80_UaM4xpVd_x7/view)

In short, there are at least two feedback mechanisms at play here.

1. `FrameStabilityContainer` is at time ~1899, and proposes to advance to ~1901. The next available frame is at time 1900 - so FSC is allowed to advance to 1900.
2. FSC is at time 1900, but it proposed to advance to ~1901, so it operates as if it is actually falling behind on a rewind and starts going backwards for a few frames.
3. At some point FSC manages to succeed after proposing a time at some point before 1900 (say, ~1890). The next frame it tries to go forward again, and is blocked on pending frames, at which point it starts going backwards again to catch up.
4. Eventually FSC succeeds in proposing a time before ~1890, and is allowed to proceed to it. Rinse and repeat this back and forth (a few frames at a time, every time), until
5. FSC rewinds past 1700 at which point there are 2 pending frames and normal playback resumes until the threshold of 1900 is crossed again.

Gonna try to rack my brain for a possible fix but at this point it might honestly be easier to just disable rewind entirely when spectating? Not sure...

bdach commented:
Still got nothing after ~2 hours. I feel like the logic of checking for new frames should be separated out of the time-dependent logic, as the one extra update loop is triggering this. Commenting it out makes the gameplay stop properly (but not resume, understandably).

I've tried fiddling with multiple things but I got nothing. The parent gameplay clock moving on its own thwarts all my efforts - even if I can get FSC to stay in place time-wise, the parent clock will shift back after hitting the last frame boundary and trigger all of the rewind logic that causes this to happen in the first place

bdach commented on osu.Game/Rulesets/UI/FrameStabilityContainer.cs:
```diff
@@ -119,93 +115,109 @@ private void updateClock()
 
-            var newProposedTime = parentGameplayClock.CurrentTime;
+            // our goal is to catch up to the time provided by the parent clock.
+            var proposedTime = parentGameplayClock.CurrentTime;
```
Actually, now that I look at it again,

```suggestion
            var proposedTime = frameStableClock.IsPaused.Value ? manualClock.CurrentTime : parentGameplayClock.CurrentTime;
```

*sorta kinda* fixes it, but I don't even know where to *begin* justifying that change...

peppy replied:
As discussed on stream, fixed this in a very similar way. Need confirmation it works for you, as I can't reproduce locally.

bdach replied:
I can confirm that the latest changes fix the oscillation issue.

peppy committed:
Move initial validity check out of loop for clarity

peppy committed:
Avoid sourcing parent clock when in a paused state

peppy committed:
Make tests more resilient under headless execution

peppy committed:
Add further test logic to ensure retry / restart flow is working correctly

peppy committed:
Fix stop watching test to check correct screen presence

peppy committed:
Add test covering host retry after returning to spectator screen

peppy committed:
Rename starvation test

peppy committed:
Move helper functions to bottom of class

peppy committed:
Move clock retrieval to new correct location

peppy committed:
Fix aim time being mutated inside update loop

peppy commented:
I'm very unsure how the failing test is failing. Been trying to repro it locally but can't over countless headless runs.

peppy committed:
Merge branch 'master' into spectator-replay-watcher

bdach commented:
I think I am able to reproduce the test failure, or at least *a* failure that is similar to it. To reproduce, comment/ignore all of the tests except for the failing one (`TestPlayStartsWithNoFrames`) and run it in the test browser. Headless seems to pass. ([Don't ask how I got this "bright" idea.](https://twitter.com/Spaceman_Atlas/status/1321863418563956736))

I will attempt to describe the failure case, with log excerpts this time, to hopefully describe better what's happening.

<details>
<summary>Failure case description (or at least an attempt of one)</summary>

FSC = `FrameStabilityContainer`
FRIH = `FramedReplayInputHandler`

```
12411: step 13 send frames
```

The "send frames" step executes. FSC's `UpdateSubTree` runs:

```
[runtime] 2020-10-29 19:54:15 [verbose]: == UPDATESUBTREE START ==
[runtime] 2020-10-29 19:54:15 [verbose]: FSC waiting on frames = True, paused = True, manual @ -1043.595439, gameplay @ -1084.595439, framed @ -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = 1, next frame @ 0, current time = , proposed time = -1043.595439, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -1043.595439
```

FSC is paused, so it sets frame from the manual clock's time. FRIH wants to advance forward correctly to the frame at time 0, but the time given in FSC is actually in the past relative to 0. Since the directions don't match, and the conditions to block are not met, it suggests to stay at the same time instant instead of advancing to the newly-received first frame.

```
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 0, manual clock @ -1043.595439 (rate 0), framed clock @ -1043.595439 (rate 0)
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = 1, next frame @ 0, current time = -1043.595439, proposed time = -1043.595439, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 0, manual clock @ -1043.595439 (rate 0), framed clock @ -1043.595439 (rate 0)
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = 1, next frame @ 0, current time = -1043.595439, proposed time = -1043.595439, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 0, manual clock @ -1043.595439 (rate 0), framed clock @ -1043.595439 (rate 0)
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = 1, next frame @ 0, current time = -1043.595439, proposed time = -1043.595439, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 0, manual clock @ -1043.595439 (rate 0), framed clock @ -1043.595439 (rate 0)
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = 1, next frame @ 0, current time = -1043.595439, proposed time = -1043.595439, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 0, manual clock @ -1043.595439 (rate 0), framed clock @ -1043.595439 (rate 0)
```

This continues for the 5 times FSC is allowed to try and catch up to the gameplay clock in one frame.

On the next frame:

```
[runtime] 2020-10-29 19:54:15 [verbose]: == UPDATESUBTREE START ==
[runtime] 2020-10-29 19:54:15 [verbose]: FSC waiting on frames = False, paused = False, manual @ -1043.595439, gameplay @ -1124.6999797887659, framed @ -1043.595439
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = -1, next frame @ 0, current time = -1043.595439, proposed time = -1060.2621056666667, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed 0
```

Since `WaitingOnFrames` has become false, the gameplay clock is still going backwards, but is no longer paused. Stability is applied, clamping -1124 to -1060 (~16ms before manual clock's time). FRIH now supposes it is going backwards (wrongly, since it has actually never even played a frame), and the directions match (also wrongly), so it proposes advancing to time 0 despite technically doing playback backwards. The playback state is valid, so the catch-up logic is allowed to run.

```
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = 1, manual clock @ 0 (rate 0.021840392924351245), framed clock @ 0 (rate 0.021840392924351245)
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH direction = -1, next frame @ , current time = 0, proposed time = -16.666666666666668, important section = False, replay done = False
[runtime] 2020-10-29 19:54:15 [verbose]: FRIH proposed -16.666666666666668
[runtime] 2020-10-29 19:54:15 [verbose]: FSC direction = -1, manual clock @ -16.666666666666668 (rate -0.021840392924351245), framed clock @ -16.666666666666668 (rate -0.021840392924351245)
```

While both clocks have now moved to the first frame correctly, both FRIH and FSC want to go backwards still, therefore FRIH blocks playback at time 0 due to apparent frame starvation.

</details>

No bright ideas on how to fix right now, sorry. Even grasping this scenario mentally in its entirety is at the limit of my cognition right now.

peppy commented:
Does that repro occur 100%? I can't seem to make it happen. The details you provided should be enough to figure out a solution, but still weird this is the second time I haven't been able to repro an issue that you can..

peppy committed:
Avoid running full updateClock loop when waiting on frames

peppy commented:
I've pushed a fix which I *think* should work to cover the case you describe (or, basically, if it used to work then it should still work after this PR). No longer running the updateClock loop in the case of waiting for frames.

I'm not sure if there's still issues with the updateClock loop in the scenario you have mentioned so be on the lookout for anything else weird.

bdach commented:
It wasn't entirely 100% (seemed to be timing dependent as breakpointing in places sometimes allowed the test to pass) but it was consistent if only logging output was tacked on, yeah.

Zero clue why you can't seem to reproduce any of these, and the test is still failing :( I guess I'll have to continue my insane exercise of trying to debug on appveyor if the fix does work for me locally...

peppy commented:
For what it's worth I can repro on windows.

bdach commented:
I just noticed the discord message where you mentioned offsets - I in fact do have a custom offset in my lazer config, and windows also has a hardcoded offset of 15ms I wanna say? That'll probably be the key piece as to why our execution states seem to be different.

peppy commented:
Oh damn. I think you may be on to something there.

peppy committed:
Fix loop not exiting after first valid frame

peppy committed:
Fix direction and IsRunning not updating on first frame after becoming valid

The parent clock will not unpause until WaitingForFrames becomes false,
so I've moved the set of that before we start to propagate its values
across. Doesn't fix any visible issue but should make propagation one
game loop faster.

peppy committed:
Fix FramedReplayInputHandler incorrectly blocking in streaming mode when time requested is before the first frame

Most of this is just tidying up the logic to (hopefully) be better to
follow, again (again (again)).

The actual fix is that we now allow interpolation/playback when the
incoming time is less than the first frame's time, regardless of
receiving status.

peppy committed:
Add test coverage for streaming replay playback

peppy committed:
Merge branch 'master' into spectator-replay-watcher

smoogipoo commented on osu.Game/Rulesets/UI/FrameStabilityContainer.cs:
```diff
@@ -85,35 +85,52 @@ protected override void LoadComplete()
+
+            double aimTime = parentGameplayClock.CurrentTime;
+
+            if (frameStableClock.WaitingOnFrames.Value)
```
Am I the only one that can't follow this logic? It's split between 3 methods - updateClock(), updateReplay() and now UpdateSubTree(), all of which directly contribute to the state.

I really don't like this. Why can't this be inside updateClock() at the very least, with a further improvement being to separate it out into its own method that isolates its purpose?

peppy replied:
can you explain which logic you mean? the fact that this variable is set inside `updateClock`?

smoogipoo replied:
All of the new logic added in UpdateSubTree is very hard to follow. Including the parent gameplay clock now being set here instead of updateClock()

peppy replied:
I can attempt to move it back. I initially moved it over to attempt to fix the reported issue, but it turns out that was not required.

smoogipoo commented on osu.Game/Rulesets/UI/FrameStabilityContainer.cs:
```diff
@@ -85,35 +85,52 @@ protected override void LoadComplete()
+
+            if (frameStableClock.WaitingOnFrames.Value)
+            {
+                // waiting on frames is a special case where we want to avoid doing any update propagation, unless new frame data has arrived.
```
What's the issue with the "update propagation" happening? All we're doing is updating the clocks right? Over which both updateReplay() and updateClock() have full control over the proposed time and can choose to set it arbitrarily...

Regardless of how that's done the final state should be NotValid and the update "propagation" won't happen anyway?

peppy replied:
I moved this out to avoid running the updateClock method while waiting on frames. By running it, it updates the manualClock's state which I wanted to avoid at all costs.

But I think you may be right in that we don't want to call SetFrameFromTime twice in a row, as it will advance two frames, potentially.

smoogipoo replied:
> we don't want to call SetFrameFromTime twice in a row

That's the sort of stuff I'm concerned about. As another example, this also has no check for `ReplayInputHandler` being null, which is already checked prior to `updateReplay()`. Which is why I mentioned that it's hard to follow this logic above.

> updates the manualClock's state which I wanted to avoid at all costs

But why? What part of it would break? Couldn't this very same logic, placed inside updateReplay(), set `proposedTime = manualClock.CurrentTime` and forego such updates?

smoogipoo reviewed (changes requested):
Can this replay stuff be split out into a separate PR?

peppy committed:
Move clock logic back to inside updateClock method

peppy committed:
Move initial state set inside updateClock

peppy committed:
Merge branch 'master' into spectator-replay-watcher

bdach committed:
Fix inaccurate xmldoc

bdach committed:
Add more xmldocs

bdach committed:
Merge branch 'master' into spectator-replay-watcher

bdach committed:
Ensure spectator player is unsubscribed to prevent leak

bdach commented:
I've gone through this code again and pushed in a few changes. Most are xmldocs, but one is pretty important (7164583) - pretty sure the lack of unsubscriptions there was keeping instances of `SpectatorPlayer` alive.

The oscillation issue [I described here](https://github.com/ppy/osu/pull/10605#issuecomment-718119011) is back, but it does seem to be caused by incorrect offset application which is now tracked separately in #10646, and this diff is large enough as it is on its own, so I'd say let's ignore for now.

I don't see much reason to hold up this PR any more, although I'm still a little cagey on the clock logic. It reads correctly to me, as in I can't spot anything that would be wrong, but there are a lot of things that *aren't* there but should be that I could be missing.

smoogipoo commented on osu.Game/Rulesets/UI/FrameStabilityContainer.cs:
```diff
@@ -85,34 +85,46 @@ protected override void LoadComplete()
                 base.UpdateSubTree();
                 UpdateSubTreeMasking(this, ScreenSpaceDrawQuad.AABBFloat);
-            }
+            } while (state == PlaybackState.RequiresCatchUp && loops-- > 0);
```
nit

```suggestion
            } while (state == PlaybackState.RequiresCatchUp && --loops > 0);
```

smoogipoo committed:
Merge branch 'master' into spectator-replay-watcher

smoogipoo reviewed (approved):
Looks good to me
